### ISSUE TITLE
(Meta)Data Validation and Document Types

### ISSUE THREAD

**Issue by rufuspollock** [datopian/markdowndb#55]
When loading a file I want to validate it against a schema/type so that I know the data/content in my contentbase/"database" is valid

- When validation fails what happens?
- Error messages should be super helpful
- Follow the principle of erroring early

When loading a file I want to allow "extra" metadata by default so that I don't get endless warnings about extra fields that are not defined for document type X.

When accessing a File I want to cast it to a proper typescript type so that I can use it from code with all the benefits of typescript.

BYOT (bring your own types) When working with markdowndb i want to create my own types ... so that when I get an object out it is cast to the right typescript type

## Acceptance

- [x] Design sketch in README of API for users
- [x] Example just simple usage from JS (no need for nextjs) e.g. like this

```
import indexFolder from markdowndb

// setup zod stuff and configure markdowndb with

files, errors = indexFolder(withZodStuff)
console.log(errors)
```

## Tasks

- [x] Sketch out a simple example concrete use case (write a short tutorial illustrating this hypothetical feature)
- [x] Research how astro does it with its zod stuff
- [x] sketch out the design before implementing
- [x] Implement 

## Design

1. Should the user define the scheme using Zod, or should we build a more intuitive way that doesn't require knowledge of external libraries?

I believe we should make the use of Zod optional, as it may be overkill for users. Most users likely want to validate if a specific front matter field is provided, rather than engaging in complex validation.

The best approach would be to allow users to:

1. Make a field required or optional.
2. Specify whether it's a string or a number.
3. Provide their own validation logic (giving users flexibility without imposing unnecessary complexity).
4. Choose whether to use Zod for validation.

**Example:**

If a user wants to validate that a field named `dates` with two or more dates is provided, the schema could be defined as follows:

```javascript
dates: {
    type: string,
    required: true,
    validate: (fileObject) => {
        if (/* field 'dates' is incorrect */) {
            return {
                status: false, // Error
                message: ""     // Error message
            };
        } else {
            return {
                status: true,  // Correct
            };
        }
    }
}
```

---

2. Should all the file schemes be defined in the config file?
Yes, it's cleaner, and we should allow the users to import schemes from other js files for organization.

---

3. Determining Files Matching a Scheme:

Option 1: Continue with `type` Property
- If no `pattern` field is provided, default to using the `type` property (e.g., `type: "post"`). This caters to simplicity for basic users and use cases.

Option 2: Introduce a `pattern` Property
- If a `pattern` field is provided, utilize it to match files based on the specified pattern (e.g., `pattern: "post/**"`). This offers flexibility for handling diverse and complex use cases.

**Consideration:**
The goal is to strike a balance between providing essential functionality and accommodating complex use cases. The defaulting to `type` for simplicity while allowing the use of a custom `pattern` enables users to tailor the file matching process to their specific needs.

---

4. Adding Computed Fields:

To incorporate computed fields into the scheme, a `compute` or `value` field can be introduced, accepting a function. This allows users to perform computations or set values dynamically.

**Example:**

Suppose you have a scenario where the `dates` field needs to be computed based on some dynamic logic. You can define it as follows:

```javascript
dates: {
    type: string,
    required: true,
    compute: (fileObject, ast) => {
        // Perform dynamic computation based on fileObject or ast
        // ...

        // Return the updated fileObject
        return fileObject;
    },
    validate: (fileObject) => {
        // Validation logic for the 'dates' field
        // ...
    }
}
```

In this example, the `compute` function takes the `fileObject` (representing the file's metadata) and `ast` (abstract syntax tree) as parameters. Users can apply custom logic within the `compute` function to dynamically calculate or set the value of the `dates` field.

This approach enhances flexibility by allowing users to include computed fields seamlessly within the overall scheme, contributing to a more versatile and adaptable file structure.

---

5. Also Maybe we can build npm packages with a pre-defined library of schemes or provide a comprehensive list of examples, as I consider schemes to be something not that simple for all users.
I think a few examples that cover what the user might want to do is enough.

### Architecture

I think we use zod here and typescript types. (cf astro approach)

### How does zod work?

1. **Defining Schemas:**
   - You create a schema using the various methods provided by Zod, such as `z.string()`, `z.number()`, `z.object()`, etc.
   - Example:
     ```typescript
     import { z } from "zod";

     const userSchema = z.object({
       username: z.string(),
       age: z.number().min(18),
     });
     ```

2. **Validation:**
   - You can then use the `parse` method to validate and parse data according to the defined schema.
   - Example:
     ```typescript
     const validUserData = userSchema.parse({
       username: "john_doe",
       age: 25,
     });
     ```

   - If the provided data doesn't match the schema, a `ZodError` is thrown with details about the validation errors.

3. **Custom Error Messages:**
   - You can customize error messages to provide meaningful feedback to users.

   ```typescript
   const userSchema = z.object({
     username: z.string().min(3, { message: "Username must be at least 3 characters" }),
     age: z.number().min(18, { message: "Age must be at least 18" }),
   });
   ```

**Comment by olayway** [comment]
@rufuspollock I've advised @mohamedsalem401 to first sketch this out before writing any code

**Comment by mohamedsalem401** [comment]
@rufuspollock 
Let's explore a straightforward example involving a content management system responsible for storing blog posts. Each blog post adheres to a specific structure, encompassing properties such as `title`, `content`, `tags`, and `author`. Our focus is on devising the file loading process, employing the previously delineated concepts.

```json
// blogPostSchema.json
{
  "pattern": "blog/*.md",
  "properties": {
    "title": { "type": "string", "required": true },
    "date": { "type": "string" },
    "author": { "type": "string", "default": "Anonymous" },
    "tags": {
      "type": "array",
      "items": { "type": "string" },
      "minItems": 1,
      "uniqueItems": true
    }
  },
  "additionalProperties": false // allow extra metadata (true by default)
}
```

In the event that a user attempts to execute the `indexFolder` function and a file fails to adhere to the specified schema, an error message like the following would be displayed:

```plaintext
Error: Title field is required on type Post at blog/post1.md
```

When it comes to comparing schemas, TypeScript proves to be a valuable tool. By converting the schema into a TypeScript interface, we can facilitate straightforward comparisons. Consider the following examples:

**Schema:**
```json
{
  "type": "object",
  "properties": {
    "title": { "type": "string", "required": true },
    "content": { "type": "string" },
    "author": { "type": "string" }
  }
}
```

**Equivalent TypeScript Interface:**
```typescript
interface BlogPost {
  title: string;
  content?: string;
  author?: string;
}
```

This approach allows us to define TypeScript interfaces mirroring the JSON schema, enabling us to harness the TypeScript type system for identifying potential type errors during development.

**Comment by rufuspollock** [comment]
@mohamedsalem401 what does zod do in astro build?

**Comment by rufuspollock** [comment]
@mohamedsalem401 i was more thinking about the use of zod in the astro content collections feature which is similar to what we have in markdowndb https://docs.astro.build/en/guides/content-collections/

**Comment by mohamedsalem401** [comment]
1. **Configuring Content Collections in `src/content/config.ts`:**
   - In the `src/content/config.ts` file, the user configures content collections for an Astro project.
   - The configuration involves specifying the type of collection and defining a schema for frontmatter validation.

2. **Using `zod` for Schema Validation:**
   - Astro utilizes `zod` directly for schema validation.
   - Users need to learn how to use `zod` to define and validate the structure of frontmatter in their content collections.

3. **Searching for Astro Configuration Files:**
   - There's a function that searches for Astro configuration files (`astro.config.js|cjs|mjs|ts`) in different formats.
   - This function operates within a specified root directory.

4. **Loading Astro Configuration with Vite:**
   - Another function loads the Astro configuration using Vite, a frontend build tool.

5. **Defining Collections in `src/content/config.ts`:**
   - Collections are defined in the `src/content/config.ts` file using the `defineCollection` function.
   - All defined collections are exported as part of the `collections` object.

6. **Querying Documents of Type "blog":**
   - To retrieve documents of type "blog," there's an example usage of the `getCollection` function.
   - The function is invoked with the argument `'blog'` to fetch all entries from the "blog" collection.
   - The result, stored in `allBlogPosts`, contains the content entries from the specified collection.

In summary, the user configures content collections in `src/content/config.ts`, defines collection types and schemas using `zod`, and utilizes Astro functions like `getCollection` to query and retrieve content entries based on the defined collections.

**Comment by mohamedsalem401** [comment]
Some of the questions that we might answer:
1. Should the user define the scheme using Zod? Or should we build a more intuitive way that doesn't require the knowledge of external libraries?

2. Should all the file schemes be defined in the config file?

3. Astro knows the files that match a file scheme by the `type` property. For example, a scheme with `type: "post"` will make all files in `content/post/**` become post files. Should we do this? Or use a `pattern` property, e.g., a scheme with `pattern: "post/**"` will make all files in `content/post/**` become post files.

4. How to add computed fields in all of this? Maybe we can do something like:

```javascript
function getContentLength(data) {
  return data.content.length;
}

scheme: {
    content: z.string(),
    length: {
      scheme: z.string().optional(),
      value: getContentLength
    }
}
```
or 
```javascript
function getContentLength(data) {
  return data.content.length;
}

scheme: {
    content: z.string(),
    length: z.string().optional().transform(getContentLength) // pass the function directly to `zod`'s `transform` method
}
```

5. Also Maybe we can build npm packages with a pre-defined library of schemes or provide a comprehensive list of examples, as I consider schemes to be something not that simple for all users.

**Comment by mohamedsalem401** [comment]
@rufuspollock

I have added answers for how I would address the document types design in the `Design` section and also explained how Zod works in the notes section.

**Comment by rufuspollock** [comment]
@mohamedsalem401 now we have a basic sketch let's start creating a really simple example where we use zod to set types and then have validate and reporting on build of markdown in some useful way. This could start as a PR an example in the examples folder. I've added stuff to acceptance for you.
